core/domain/exp_services.py

```diff
@@ -2063,6 +2063,345 @@ def get_interaction_id_for_state(exp_id, state_name):
         'There exist no state in the exploration with the given state name.')
 
 
+def update_logged_out_user_progress(
+    exploration_id: str,
+    unique_progress_url_id: str,
+    state_name: str,
+    exp_version: int,
+    ) -> None:
+    """Updates the logged-out user's progress in the
+        associated TransientCheckpointUrlModel.
+
+    Args:
+        exploration_id: str. The ID of the exploration.
+        unique_progress_url_id: str. Unique 6-digit url to track a
+            logged-out user's progress.
+        state_name: str. State name of the most recently
+            reached checkpoint in the exploration.
+        exp_version: int. Exploration version in which a
+            checkpoint was most recently reached.
+    """
+    # Fetch the model associated with the unique_progress_url_id.
+    model = exp_models.TransientCheckpointUrlModel.get(unique_progress_url_id)
+
+    # Create a model if it doesn't already exist.
+    if model is None:
+        model = exp_models.TransientCheckpointUrlModel.create(
+            exploration_id, unique_progress_url_id)
+
+    current_exploration = exp_fetchers.get_exploration_by_id(
+        exploration_id, True, exp_version)
+
+    # If the exploration is being visited the first time.
+    if model.furthest_reached_checkpoint_state_name is None:
+        model.furthest_reached_checkpoint_exp_version = exp_version
+        model.furthest_reached_checkpoint_state_name = state_name
+    elif model.furthest_reached_checkpoint_exp_version < exp_version:
```

Comment:
What about the case with same exp version ?

Reply:
Fixed.